### PR TITLE
Added support for fallback drivers

### PR DESCRIPTION
**Pull request by VuongN**
Thanks for the great work @MikaAK.  I needed the feature for my project and thought I'd contribute back. Perhaps it will help others as well.

- Added support for fallback drivers in order of preference (https://github.com/MikaAK/angular2-locker/issues/16) which allows user to pass in a single `Driver` or an `Array<Driver>`.
- Added unit tests coverage
- Modified documentation
- I tried really hard to not disturb the current code, but alas, my OCD instinct took over and I had to make some minor changes to files that I touched (added semi-colons & {} to blocks etc.).  I could perhaps issue another PR to fix all minor code quality issue if you deem it necessary, @MikaAK.

### CODE REVIEW

**Comment by MikaAK** [comment]
Thanks for the PR, this feature is greatly needed! I'll merge this into my newest version that a few fixes! Yeah, I'm a semicolon less person haha, I've never used them in 3 years so it's just habit now! If you could fix it up and remove them and then run a `git commit -n --amend` then a force push up to your branch it would let me merge this in without additional commits :)

**Comment by VuongN** [comment]
Heh. Alright, I removed the semicolons.  It is against what I advocate for in term of good development habit. But since you own the project, it's your call.  Let me know if you need anything else.

**Comment by MikaAK** [comment]
Would you mind rebasing the work with latest master? Lots of new changes in! After that I'll merge it in :D

File: src/Driver.ts [hunk 1]
```diff
@@ -42,8 +42,7 @@ export class Driver {
       this.get(LOCKER_TEST_KEY)
       this.remove(LOCKER_TEST_KEY)
     } catch (e) {
-      if (__DEV__)
-        console.error(e)
+      __DEV__ && console.error(e)
```
Comment (MikaAK) [inline src/Driver.ts]: `&&` shouldn't be used in this case as it's better suited for boolean checks not control flow. Prefer if/else for control flow

File: src/Driver.ts [hunk 2]
```diff
@@ -58,3 +57,21 @@ export const DRIVERS = {
   MEMORY: new Driver(new MemoryStorage()),
   COOKIE: new Driver(new CookieStorage())
 }
+
+export function determineDriver(defaultDriverType?: any): Driver {
+  let hasFallback = Array.isArray(defaultDriverType)
```
Comment (MikaAK) [inline src/Driver.ts]: This removes ability to set single default driver type. Ideally we preserve both

File: src/Driver.ts [hunk 3]
```diff
@@ -58,3 +57,21 @@ export const DRIVERS = {
...
+
+  if (!hasFallback) {
+    return defaultDriverType.isSupported() ? defaultDriverType: DRIVERS.MEMORY
+  }
+
+  let driver = DRIVERS.MEMORY
+  for (let _driver of defaultDriverType) {
```
Comment (MikaAK) [inline src/Driver.ts]: Personally I would stick to higher order functions and do
```
const supportedDriver = defaultDriverType
  .find(driver => driver.isSupported())

if (supportedDriver)
  driver = supportedDriver
```

File: src/Driver.ts (hunk 2 above)
Comment (VuongN) [inline src/Driver.ts]: If you check function/tests again, you will see that it preserves both single and multiple.  I made sure the change is backward compatible.

**Comment by VuongN** [comment]
The name change surprised me a bit. I also noticed that unit tests broke in master. -V.

**Pull request merged**